test/unit/connection_string.test.ts

```diff
@@ -312,4 +325,107 @@ describe('Connection String', function () {
       expect(options.srvHost).to.equal('test1.test.build.10gen.cc');
     });
   });
+
+  describe('resolveSRVRecord()', () => {
+    const resolveSRVRecordAsync = promisify(resolveSRVRecord);
+
+    afterEach(async () => {
+      sinon.restore();
+    });
+
+    function makeStub(txtRecord: string) {
+      const mockAddress = [
+        {
+          name: 'localhost.test.mock.test.build.10gen.cc',
+          port: 2017,
+          weight: 0,
+          priority: 0
+        }
+      ];
+
+      const mockRecord: string[][] = [[txtRecord]];
+
+      // first call is for stubbing resolveSrv
+      // second call is for stubbing resolveTxt
+      sinon.stub(dns, 'resolveSrv').callsFake((address, callback) => {
+        return process.nextTick(callback, null, mockAddress);
+      });
+
+      sinon.stub(dns, 'resolveTxt').callsFake((address, whatWeTest) => {
+        whatWeTest(null, mockRecord);
+      });
+    }
+
+    for (const mechanism of $EXTERNAL_AUTH_SOURCE_MECHANISMS) {
+      it(`should set authSource to $external for ${mechanism} external mechanism`, async function () {
+        makeStub('authSource=thisShouldNotBeAuthSource');
+        const credentials = new MongoCredentials({
+          source: '$external',
+          mechanism,
+          username: 'username',
+          password: mechanism === AuthMechanism.MONGODB_X509 ? undefined : 'password',
+          mechanismProperties: {}
+        });
+        credentials.validate();
+
+        const options = {
+          credentials,
+          srvHost: 'test.mock.test.build.10gen.cc',
+          srvServiceName: 'mongodb',
+          userSpecifiedAuthSource: false
+        } as MongoOptions;
+
+        await resolveSRVRecordAsync(options);
+        expect(options.credentials === credentials).to.be.true;
```

Comment:
why not just check `expect(options.credentials).to.equal(credentials)`?

Reply:
This was my onion 🧅 , with my recent revelation that `.eql` means deep, and `.equal` means shallow. I have a preference to make clear our intention. We prefer `expect(x).to.deep.equal(y);` since it makes it obvious deepness is desired but there's no equivalent `.shallow`. (maybe we could add it!)

It's a rare case that shallow equality is important to a test that's checking something with objects so it feels like something that's easily overlooked, or something that when read later is thought to be a mistake. If someone changes this to be deep, it would also pass but it would lose the meaning originally intended. 

Maybe this is what code comments where invented for? 🤔

Reply:
yeah, overthinking it here a bit; `to.equal` is the correct assertion, and yes, code comments is where it's at if this is a concern, in fact, a code comment would be appropriate either way because it took me a minute to try to figure out why this assertion was there in the first place, so something like `// check reference identity because ...` would be ideal. Of course I'm sure @durran would tell you to just use a separate `it` assertion altogether to clarify the intention of your test cases; I don't have equally strong feelings about one assertion per `it` block, but I do want clarity, and using the most appropriate assertion for the job (in this case, it would actually be something like `expect(options).property('credentials').to.equal(credentials)`)